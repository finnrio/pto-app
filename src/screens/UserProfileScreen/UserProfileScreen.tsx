import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import { Alert, Text, TextInput, TouchableOpacity, View } from "react-native";
import React, { useContext, useEffect, useState } from "react";
import { updateEmail } from "firebase/auth";

import { UserContext } from "../../context/UserContext";
import GetCurrentUserData from "../../firebase/firestore/GetCurrentUserData";
import SetUserProfileData from "../../firebase/firestore/SetUserProfileData";
import styles from "./styles";
import { FIREBASE_AUTH } from "../../firebase/firebaseConfig";
import { AppUser } from "../../types/AppUser";

export default function UserProfileScreen() {
  const currentUser = useContext(UserContext);
  const [firstName, setFirstName] = useState("");
  const [surname, setSurname] = useState("");
  const [email, setEmail] = useState("");
  const [role, setRole] = useState("");

  async function RenderUserData(/* id: any */) {
    const res: AppUser = await GetCurrentUserData(); // this should use an ID if used by admin
    setFirstName(res?.first_name);
    setSurname(res?.surname);
    setEmail(res?.email);
    setRole(res?.role);
  }

  function createUserDataObject() {
    return {
      first_name: firstName,
      surname,
      email,
    };
  }

  function HandleUpdateBtn() {
    try {
      Alert.alert(
        "Update users profile details",
        "Are you sure you want to update your profile details?",
        [
          {
            text: "Update",
            onPress: () => {
              SetUserProfileData(createUserDataObject(), currentUser?.uid);
              if (email !== FIREBASE_AUTH.currentUser?.email) {
                updateEmail(currentUser!, email!).catch((error) =>
                  Alert.alert("Error", error.code),
                );
              }
              RenderUserData();
            },
            style: "default",
          },
          {
            text: "Reset Changes",
            onPress: () => RenderUserData(),
            style: "cancel",
          },
          {
            text: "Cancel",
            style: "cancel",
          },
        ],
        { cancelable: true },
      );
    } catch (e: unknown) {
      console.error(e);
      if (e instanceof Error) {
        Alert.alert(e.name, e.message);
      } else {
        Alert.alert("Error", "An error has occured");
      }
    }
  }

  useEffect(() => {
    RenderUserData();
  }, []);

  return (
    <View style={styles.container}>
      <KeyboardAwareScrollView
        style={{ flex: 1, width: "100%" }}
        keyboardShouldPersistTaps="always"
      >
        <Text style={styles.text}>User ID</Text>
        <TextInput
          style={styles.input}
          placeholder="Failed to load User ID"
          placeholderTextColor="#aaaaaa"
          value={currentUser?.uid}
          underlineColorAndroid="transparent"
          editable={false}
        />
        <Text style={styles.text}>First Name</Text>
        <TextInput
          style={styles.input}
          placeholder="First Name"
          placeholderTextColor="#aaaaaa"
          onChangeText={(text) => {
            setFirstName(text);
          }}
          value={firstName}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
        />
        <Text style={styles.text}>Surname</Text>
        <TextInput
          style={styles.input}
          placeholder="Surname"
          placeholderTextColor="#aaaaaa"
          onChangeText={(text) => {
            setSurname(text);
          }}
          value={surname}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
        />
        <Text style={styles.text}>Email</Text>
        <TextInput
          style={styles.input}
          placeholder="E-mail"
          placeholderTextColor="#aaaaaa"
          onChangeText={(text) => {
            setEmail(text);
          }}
          value={email!}
          underlineColorAndroid="transparent"
          autoCapitalize="none"
        />
        <Text style={styles.text}>Role</Text>
        <TextInput
          style={styles.input}
          placeholder="User Role"
          placeholderTextColor="#aaaaaa"
          value={role?.toString()}
          underlineColorAndroid="transparent"
          editable={false}
        />
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            HandleUpdateBtn();
          }}
        >
          <Text style={styles.buttonTitle}>Update Profile</Text>
        </TouchableOpacity>
      </KeyboardAwareScrollView>
    </View>
  );
}
