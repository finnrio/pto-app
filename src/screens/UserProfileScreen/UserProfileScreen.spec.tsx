import React from "react";
import { render } from "@testing-library/react-native";
import UserProfileScreen from "./UserProfileScreen";

// solution from https://github.com/APSL/react-native-keyboard-aware-scroll-view/issues/493#issuecomment-1023551697
jest.mock("react-native-keyboard-aware-scroll-view", () => ({
  KeyboardAwareScrollView: (props: { children: React.ReactNode }) =>
    props.children,
}));

// solution from https://github.com/expo/expo/issues/21434#issuecomment-1450781966
jest.mock("expo-font");

describe("UserProfileScreen", () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  it("renders correctly", () => {
    const tree = render(<UserProfileScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
