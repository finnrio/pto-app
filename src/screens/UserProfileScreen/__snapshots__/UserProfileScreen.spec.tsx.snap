// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfileScreen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    User ID
  </Text>
  <TextInput
    editable={false}
    placeholder="Failed to load User ID"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    underlineColorAndroid="transparent"
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    First Name
  </Text>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="First Name"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    Surname
  </Text>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="Surname"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    Email
  </Text>
  <TextInput
    autoCapitalize="none"
    onChangeText={[Function]}
    placeholder="E-mail"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text
    style={
      {
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "paddingLeft": 16,
      }
    }
  >
    Role
  </Text>
  <TextInput
    editable={false}
    placeholder="User Role"
    placeholderTextColor="#aaaaaa"
    style={
      {
        "backgroundColor": "white",
        "borderRadius": 5,
        "height": 48,
        "marginBottom": 10,
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 10,
        "overflow": "hidden",
        "paddingLeft": 16,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#788eec",
        "borderRadius": 5,
        "height": 48,
        "justifyContent": "center",
        "marginLeft": 30,
        "marginRight": 30,
        "marginTop": 20,
        "opacity": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "white",
          "fontSize": 16,
          "fontWeight": "bold",
        }
      }
    >
      Update Profile
    </Text>
  </View>
</View>
`;
